### PR TITLE
Quant: add weight int4pack mm kernel

### PR DESCRIPTION
**Pull request by yanboliang**
Adding the weight int4pack mm CUDA kernel. The kernel comes from the tinnygemm project which developed by Jeff Johnson.

### CODE REVIEW

**Review by Chillee — APPROVED** [review]
Add a brief torch.compile unit test

File: aten/src/ATen/native/cuda/int4mm.cu [hunk 1]
```diff
@@ -0,0 +1,1020 @@
...
+#include <cuda_fp16.h>
+#include <cuda_runtime.h>
+#include <mma.h>
+#endif
+#include <ATen/ATen.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/DeviceGuard.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <torch/types.h>
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/int4mm.cu:12]: you need to include something like `#include <ATen/native/cuda/_weight_int4pack_mm_impl.cuh>` and `_convert_weight_to_int4pack.cuh` or something like that

Comment (yanboliang) [reply aten/src/ATen/native/cuda/int4mm.cu:12]: ```_weight_int4pack_mm``` and ```_convert_weight_to_int4pack``` are the two functions I added in this file, I don't have these two ```cuh``` file.

Comment (yanboliang) [reply aten/src/ATen/native/cuda/int4mm.cu:12]: Should I make two fake ```cuh``` file to make CI happy?

File: aten/src/ATen/native/cuda/int4mm.cu [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+  }
+}
+
+#endif
+
+
+torch::Tensor _weight_int4pack_mm_cuda(
+    const torch::Tensor& A,
+    const torch::Tensor& B,
+    long qGroupSize,
```
Comment (malfet) [inline aten/src/ATen/native/cuda/int4mm.cu:783]: `sizeof(long) == 4` on Win64, but `8` on all other OSes. `sizeof(int64_t)` is unambiguously `8`
```suggestion
    int64_t qGroupSize,
```

File: aten/src/ATen/native/cuda/int4mm.cu [hunk 3]
```diff
@@ -0,0 +1,1020 @@
...
+#endif
+  TORCH_CHECK(false, "_weight_int4pack_mm_cuda is not available for build.")
+  return C_final;
+}
+
+// input is [n][k] (int32 dtype)
+// output is [n / 8][k / (InnerKTiles * 16)][32][innerKTiles / 2]
+torch::Tensor _convert_weight_to_int4pack_cuda(
+    const torch::Tensor& in,
+    long innerKTiles) {
```
Comment (malfet) [inline aten/src/ATen/native/cuda/int4mm.cu:963]: long is 32-bit type on Windows, but 64-bit one everywhere else
```suggestion
    int64_t innerKTiles) {
```

**Pull request closed without merging**